### PR TITLE
Use astropy and matplotlib nightly feeds for devdeps CI

### LINKED ISSUE #6019: Use Matplotlib nightly wheels in devdeps testing
### Describe the feature

I think Matplotlib now have nightly wheels: https://github.com/matplotlib/matplotlib/pull/22733#issuecomment-1085269320. I haven't read the details, but thought I'd leave a note here that it would be good to use the nightly wheels to save a bit of time (and energy!) in our `devdeps` CI by using these instead of compiling from source.

### Proposed solution

_No response_

### PR DESCRIPTION
**Pull request by ConorMacBride**
This PR updates the tox configuration to search for the latest pre-release from the Astropy and Matplotlib nightly PyPI repositories when installing `astropy` and `matplotlib` in any of the `-devdeps` environments. This should significantly speed up the devdeps cron test on GitHub Actions and the devdeps figure tests on CircleCI. The remaining dev packages (`asdf` and `mpl-animators`) seem to be pure python so should't be a significant bottleneck.

The Astropy feed is at: https://dev.azure.com/astropy-project/astropy/_artifacts/feed/nightly/PyPI/astropy/5.1.dev781/overview

The Matplotlib feed is at: https://anaconda.org/scipy-wheels-nightly/matplotlib

Closes #6019

### CODE REVIEW

**Review by nabobalis — APPROVED** [review]

**Comment by ConorMacBride** [comment]
Will check the errors in more detail tomorrow, but the correct pre-release astropy and matplotlib versions have been installed: https://output.circle-artifacts.com/output/job/3acfb9fb-115d-41dd-993c-32fc68ac3d2f/artifacts/0/.tmp/py39-figure-devdeps/figure_test_images/figure_version_info.txt

**Comment by nabobalis** [comment]
If its a good speed up, is it time to run the devdeps on a PR now?

File: tox.ini
Comment (nabobalis) [inline tox.ini:37]: The only issue I can imagine is that it might pull in more devdeps than we expect from pypi?

Comment (ConorMacBride) [reply tox.ini:37]: This tox option just adds `--pre` to the pip command. I'd need to check if that applies to the dependencies of explicitly requested packages also. It only has an effect when installing the tox reps and definitely has no effect when installing sunpy.

**Comment by ConorMacBride** [comment]
Also note that when we request the git repository directly we are guaranteed to install the latest version. If astropy or matplotlib were to stop publishing to the URLs added to tox.ini, the sunpy devdeps tests would always run against the last published nightly release and eventually start running against the latest release on the default PyPI repository, and no error would be produced.

File: tox.ini
Comment (dstansby) [reply tox.ini:37]: > The only issue I can imagine is that it might pull in more devdeps than we expect from pypi?

This might be a good thing? Perhaps we can see how it goes for a bit, and we get more errors than expected have a re-think?

**Comment by dstansby** [comment]
There doesn't seem to be a GH actions devdeps build - am I missing something?

**Comment by nabobalis** [comment]
> There doesn't seem to be a GH actions devdeps build - am I missing something?

Needs to be added to the PR CI, for now we have the figure devdeps one.

File: tox.ini
Comment (ConorMacBride) [reply tox.ini:37]: It is installing pre-releases of astropy and matplotlib dependents: `greenlet`, `numpy`, `pandas` and `scipy`. And the pre-release `numpy` seems to be causing the test fails.

Comment (nabobalis) [reply tox.ini:37]: If its a pre-release numpy causing the issue, I feel like we should inform someone upstream?

File: tox.ini
Comment (nabobalis) [inline tox.ini:38]: That is some magic.

Comment (ConorMacBride) [reply tox.ini:38]: Yeah! Found here: https://pip.pypa.io/en/stable/cli/pip_install/#pre-release-versions

**Review by nabobalis — APPROVED** [review]

**Comment by ConorMacBride** [comment]
Given https://github.com/sunpy/sunpy/pull/5939#discussion_r833466590, maybe @Cadair should review before merging? The devdeps test only takes 5 min now, and in comparison to the py310 windows test which takes 15 min, seems fine to run in a PR. (Although we should look into what windows wheels are missing.)

**Comment by nabobalis** [comment]
> Given https://github.com/sunpy/sunpy/pull/5939#discussion_r833466590, maybe @Cadair should review before merging? The devdeps test only takes 5 min now, and in comparison to the py310 windows test which takes 15 min, seems fine to run in a PR. (Although we should look into what windows wheels are missing.)

What if we swap them around, that could be faster?

**Comment by ConorMacBride** [comment]
Yeah, should probably move py310 testing to the cron stage. When we remove the scikit-image pin we'll have py310 wheels and no longer have to build it from source. Also, `astropy-healpix` and `reproject` have platform specific py39 wheels but none for py310 yet, although they are probably not slowing things down?

**Comment by nabobalis** [comment]
I was wondering if we just switched to a linux 310 job and ran that on a PR instead of having a cron stage for just that?

**Comment by ConorMacBride** [comment]
The pinned scikit-image doesn't have py310 linux wheels either.

**Comment by nabobalis** [comment]
> As the person who maintains the Astropy nightly wheels, this is my major concern. We would have no way of knowing that the version of astropy we are testing against was suddenly a month old...

Are they not updated automatically?

**Comment by Cadair** [comment]
Yeah, but if something goes wrong in the build pipeline or whatever then it can break and nobody (me) notices for a while.

File: .github/workflows/ci.yml
Comment (Cadair) [inline .github/workflows/ci.yml:48]: I still maintain we don't need or want to be running this on PRs. For the majority of contributions it's not a relevant test.

File: .github/workflows/ci.yml
Comment (Cadair) [inline .github/workflows/ci.yml:88]:
```suggestion
        - linux: py310
```

Windows is slower anyway skimage issues aside.

**Review by Cadair — APPROVED** [review]

File: .github/workflows/ci.yml
Comment (nabobalis) [reply .github/workflows/ci.yml:48]: I will fix this in my open PR

File: .github/workflows/ci.yml
Comment (nabobalis) [reply .github/workflows/ci.yml:88]: I will fix this in my open PR

**Pull request merged**
